jep/302/README.adoc

```diff
@@ -0,0 +1,260 @@
+
+Git offers in this matter a powerful user-space tool that allows us to version,
+and quickly rollback to some previous state if need be.
+
+Git is also a very common tool nowadays for developers,
+hence it will help making Essentials more accessible to contributors.
+
+== Why not use compatibleSinceVersion metadata
+
+TIP: A given plugin can indicate what is "the oldest version [...] configuration-compatible with.".
+In other words, if I am changing the configuration format of a plugin in non backwards compatible manner, I am _expected_ to express it as such using link:https://jenkinsci.github.io/maven-hpi-plugin/hpi-mojo.html#compatibleSinceVersion[`compatibleSinceVersion`].
+
+So, this means we could use this field to _optimize_ the way we roll back: once a configuration file has been upgraded by a plugin update, we can exclude it from rollback if the `compatibleSinceVersion` field was not changed.
+
+We decided to not do, for now at least, for two reasons:
+
+* this looks like an _optimization_. $
+Hence as such, this is probably premature to try and be very smart with the way the downgrade will work ;
+* there is no guarantee that all plugin developers currently know and think about updating this metadata when needed.
+This is likely something that will be improveable when we work on link:https://issues.jenkins-ci.org/browse/JENKINS-49806[JENKINS-49806] and write the associated Jenkins Enhancement Proposal.
+
+== Backwards Compatibility
+
+There are no backwards compatibility concerns related to this proposal.
+
+== Security
+
+=== Secrets
+
+Versioning secrets is not an issue, as the snapshotting system is going to be local to the running instance.
+As the Git repository data will never be pushed _outside_, no data leak is expected from this side.
+
+=== Man In The Middle
+
+The main issue here is that an attacker could for instance instruct the evergreen client to ignore everything (by putting `*` in `.gitignore`), hence make it impossible to rollback.
+
+But this would mean someone was able to talk with connected instances.
+So even if this is a valid concern, this is considered a larger scope issue that will be addressed through link:https://issues.jenkins-ci.org/browse/JENKINS-49844[JENKINS-49844].
+
+Hence there are no *specific* security risks related to this proposal.
+
+== Infrastructure Requirements
+
+There are no new infrastructure requirements related to this proposal.
+
+== Testing
+
+We must create an image of _Essentials_ preconfigured with a complete set of representative data.
+
+Creating/defining this data clearly requires human work, but the following checks are deemed automatable.
+
+=== Upgrading/downgrading
+
+Before delivering updates on real connected instances, we must test at least the following scenarios.
+
+* Apply the upgrade or downgrade, then check the instance is _running fine_:
+** No administrative monitor triggered
+** No new ERROR logs showing up
```

Comment:
You mean, `Level.ERROR`? Hardly useful: few plugins log at this level to begin with. Almost everything of interest would be a `WARNING`.

Reply:
Yes and no. But your point makes me realize I really need to _externalize_ this concept. We need a standalone design for determining Jenkins health I think. It's a critical condition/piece for many components: for testing, for evergreen-client to decide to rollback or not, etc. And it is not (necessarily) related to the data snapshotting system.

So will refactor to make this generic and point to something to be done in this regard.

Also commented in https://issues.jenkins-ci.org/browse/JENKINS-49406?focusedCommentId=332233&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-332233 about that